app/models/messaging/bsnl/dlt_template.rb

```diff
@@ -23,11 +26,31 @@ def initialize(dlt_template_name)
     @status = STATUSES[details["Template_Status"]]
     @is_unicode = details["Is_Unicode"]
     @keys = details["Template_Keys"]
+    @version = details["Version"]
+    @is_latest_version = details["Is_Latest_Version"]
     @max_length_permitted = details["Max_Length_Permitted"].to_i
     @non_variable_text_length = details["Non_Variable_Text_Length"].to_i
     @variable_length_permitted = max_length_permitted - non_variable_text_length
   end
 
+  def self.latest_name_of(dlt_template_name)
+    BSNL_TEMPLATES.dig(drop_version_number(dlt_template_name), "Latest_Template_Version")
+  end
+
+  def self.drop_version_number(dlt_template_name)
+    dlt_template_name.chomp(".#{version_number(dlt_template_name)}")
+  end
+
+  def self.version_number(dlt_template_name)
+    version = dlt_template_name.split(".").last
+
+    if numeric?(version)
+      version.to_i
+    else
+      DEFAULT_VERSION_NUMBER
+    end
+  end
```

Comment:
These mechanics are a bit awkward because we've got the version number in the template name, yes, but also as a key in the template. Looks like we _also_ have a reference to the latest version in each blob, which will have to be updated for every existing version of a template if we update a template. Could I suggest simplifying the data model a little bit and making these mechanics methods a bit smarter?

Eg.

```yml
template_name.thing.1: # we can keep the suffix optional of course, it's just a way to give each template a unique name in this YML file
  version: 1
template_name.thing.2:
  version: 2
```

```ruby
def versions(template_name)
  templates.select { |k, v| k.start_with?(template_name) }
end

def latest_version(template_name)
  name, template = versions(template_name).max_by {|k, v| v["version"] || DEFAULT_VERSION_NUMBER}
  # return whatever you want. the name or the template
end
```

This way:
* We don't have to add and maintain a latest version ref to each template
* We don't have to add a version key to templates with only a single version
* We can rely on the version key, and not worry about the template names so much

Reply:
I've chosen to add the references to the YAML since doing the `max_by` requires us to iterate over all the templates to find the latest at "runtime". We'll make 40-50k calls to `DltTemplate.new` every day, at which point the performance gain by storing the version in the YAML once would be. The tradeoff

> which will have to be updated for every existing version of a template if we update a template

This should be ok since it's automated. One shouldn't have to touch the YAML directly ever.

Reply:
I've chosen to add the references to the YAML since doing `max_by` requires us to iterate over all the templates to find the latest at "runtime". We'll make 40-50k calls to `DltTemplate.new` every day, so it's more performant to gain by storing the version in the YAML once would be. I haven't quantified what the gains would be (we have ~200 templates) but can check that out if the tradeoff seems too bad.

> which will have to be updated for every existing version of a template if we update a template

This should be ok since it's automated. One shouldn't have to touch the YAML directly ever.

Reply:
Ohhh I see. Yeah if that thing is auto-generated, then sounds good!